## Problem 21: Amicable Numbers

Let $d(n)$ be defined as the sum of proper divisors of $n$ (numbers less than $n$ which divide evenly into $n$).
If $d(a) = b$ and $d(b) = a$, where $a \neq b$, then $a$ and $b$ are an amicable pair and each of $a$ and $b$ are called amicable numbers.

For example, the proper divisors of $220$ are $1, 2, 4, 5, 10, 11, 20, 22, 44, 55$ and $110$; therefore $d(220) = 284$. The proper divisors of $284$ are $1, 2, 4, 71$ and $142$; so $d(284) = 220$.

Evaluate the sum of all the amicable numbers under $10000$.

In [1]:
import numpy as np

In [2]:
''' 
Finds the divisors of a particular number. 
Only need to search up to the square root of the number.

eg:
>>> divisor_list(12)
>>> [1, 2, 6, 3, 4]
'''

def divisor_list(number : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Finds the divisors of a particular number. 
        Only need to search up to the square root of the number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : number : The number to find the divisors for.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : divisors : The list of divisors.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> divisor_list(12)
        >>> [1, 2, 6, 3, 4]
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(number, int): raise ValueError('Please enter an integer > 0 for the number argument.')
    if number < 0: raise ValueError('Please enter an integer > 0 for the number argument.')
    
    if number == 0: return [0]
    
    divisors = []
    for i in range(1, int(np.floor(np.sqrt(number)))+1):
        if number % i == 0: 
            divisors.append(i)
            divisors.append(int(number / i))
    
    if number in divisors: divisors.remove(number)
    if number % np.sqrt(number) == 0: divisors.remove(int(np.sqrt(number)))
    
    return divisors

In [3]:
''' 
Creates a list of the sum of the proper divisors for all number below max_number.

eg:
>>> divisor_sum_list(10)
>>> [0, 1, 1, 3, 1, 6, 1, 7, 4]
'''

def divisor_sum_list(max_number : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Creates a list of the sum of the proper divisors for all number below max_number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : max_number : The max number to calculate the divisor sum for.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : divisor_sums : The list of divisor sums.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> divisor_sum_list(10)
        >>> [0, 1, 1, 3, 1, 6, 1, 7, 4]
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(max_number, int): raise ValueError('Please enter an integer > 0 for the max_number argument.')
    if max_number < 0: raise ValueError('Please enter an integer > 0 for the max_number argument.')
    
    divisor_sums = []
    
    for number in range(max_number):
        number_divisor_list = divisor_list(number)
        divisor_sums.append(sum(number_divisor_list))
    
    return divisor_sums

In [4]:
''' 
Finds the amicable numbers up to max_number.

eg:
>>> amicable_numbers_list(285)
>>> [220, 284]
'''

def amicable_numbers_list(max_number : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Finds the amicable numbers up to max_number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : max_number : The max number to calculate the divisor sum for.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : amicable_numbers : The list of amicable numbers.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> amicable_numbers_list(285)
        >>> [220, 284]
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(max_number, int): raise ValueError('Please enter an integer > 0 for the max_number argument.')
    if max_number < 0: raise ValueError('Please enter an integer > 0 for the max_number argument.')
    
    divisor_sums = divisor_sum_list(max_number)
    
    amicable_numbers = []
    for i in range(max_number):
        if divisor_sums[i] < max_number:
            if i != divisor_sums[i]:
                if i == divisor_sums[divisor_sums[i]]:
                    amicable_numbers.append(i)
    
    return amicable_numbers

In [5]:
''' 
Finds the sum of all amicable numbers up to max_number.

eg:
>>> amicable_numbers_sum(285)
>>> (2, 504, [220, 284])
'''

def amicable_numbers_sum(max_number : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Finds the sum of all amicable numbers up to max_number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : max_number : The max number to calculate the divisor sum for.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : The following information is returned:
            : int : The number of amicable numbers.
            : int : The sum of the amicable numbers.
            : list : The list of amicable numbers.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> amicable_numbers_sum(285)
        >>> (2, 504, [220, 284])
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(max_number, int): raise ValueError('Please enter an integer > 0 for the max_number argument.')
    if max_number < 0: raise ValueError('Please enter an integer > 0 for the max_number argument.')
    
    amicable_numbers = amicable_numbers_list(max_number)
    
    return len(amicable_numbers), sum(amicable_numbers), amicable_numbers

In [6]:
problem_number = 10_000

solution = amicable_numbers_sum(problem_number)

print(f'There are {solution[0]} amicable numbers below {problem_number:,} and their sum is {solution[1]:,}.')
print(f'These numbers are {solution[2]}.')

There are 10 amicable numbers below 10,000 and their sum is 31,626.
These numbers are [220, 284, 1184, 1210, 2620, 2924, 5020, 5564, 6232, 6368].


### Problem 21 Solution:

There are 10 amicable numbers below 10,000 and their sum is 31,626.

These numbers are [220, 284, 1184, 1210, 2620, 2924, 5020, 5564, 6232, 6368].